### PR TITLE
Adding the Multilanguage Associations Manager

### PR DESCRIPTION
**Pull request by rdeutz**
# Executive summary
This PR is a partial rewrite of the com_associations component based ion the work done in the Google Summer of Code Project "Improved Multilingual" (Repo: https://github.com/joomla-projects/gsoc16_improved-multi-lingual)

## Background information
The original concept made a lot of assumptions about fields and extensions and instantiated extensions models to get information about the internal structure. This concept comes with two problems:

1. Guessing can be wrong :-)
2. Instantiate a model can end up in fatal errors and a not working Website when the extension has a bug.  

I removed the guess and model instantiating and replaced it with a help class for extensions. This helper class provides all the needed information.

# Backwards compatibility
It is a new function so no problem here

# Translation impact
Two new language files for com_associations and one new language for any core extension that supports associations.

# Testing instruction 

* clone or get a copy of the staging branch
* setup a multilingual website
* Apply patch
* discover install

You will now find a new extension under components. 
The select box allows to chose an item type.

ml-pic1

When you have chosen a type and language, you can see your articles. 

ml-pic2

Click on an article (reference) and you will see the reference article on the left side. On the right side you can chose a translation target and translate your article.

ml-pic3

If you click on the available translation you get directly the view with two articles side by side.

ml-pic4

# Developer documentation
If your extension support association and you will make your extension available in the Associations Manager you have to provide a helper class (we have examples for com_content, com_contact, com_newsfeeds, com_menus, please look at the example implementation for inspiration :-))

You need to create a file "associations.php" in your component admin helper directory. If your component is names foobar the help class must be named as "FoobarAssociationsHelper" and it must extend "JAssociationExtensionHelper". 

Define three protected vars:

$extension = 'com_foobar';
$itemTypes = array('your_itemname', 'category');
$associationsSupport = true;

Define three public functions:

public function getAssociations($typeName, $id)

This function provides an interface the "JLanguageAssociations::getAssociations" and returns an array of associated items 

public function getItem($typeName, $id)

This function should return a JTable Object for the type and id. Don't forget something like "JTable::addIncludePath(__DIR__ . '/../tables')" so that the class can be found 

public function getType($typeName = '')

Here you set up information per type, the function returns an array: 

array(
			'fields'  => $fields,
			'support' => $support,
			'tables'  => $tables,
			'joins'   => $joins,
			'title'   => $title
		);

For fields and support we have a template function that set defaults so the you only have to overwrite what is different in your extension. Make sure that $fields['title'] and $fields['state'] are set right.

We need to write some documentation in JDocs but I think that here should be enough as a start.

# Open Issues
* Ordering not great
* Deal with checkout items  
* Extensions with multiple use of categories (seems a theoretical problem, because we couldn't find an extension using it)

# Final words
I would like to says thanks to the GSOC project members you did most of the work.

### CODE REVIEW

**Comment by infograf768** [comment]
We do need indeed some inpt to solve the checkedout.
And also the ACL: do we let edit/create associations for items where one has no create/edit ACL, on individual as well as extension level.

**Comment by laoneo** [comment]
This is a very handy functionality, but should this not being built into the article manager itself, when editing an article? On a first glance it will be difficult to understand why the compare editor is on a completely different place, despite that the association is managed in the article itself.

**Comment by infograf768** [comment]
this does not only deals with articles. it is a general interface concerning all associations. here already implemented for core extensions, i.e. articles, menu items, contacts, newsfeeds, and will be for weblinks. we already can deal, article per article, generally speaking with all core extensions, item per item. com_associations is centralising all these (including 3pd if they set it) in a unique interface.

**Comment by infograf768** [comment]
@rdeutz 
please add as new feature

**Comment by laoneo** [comment]
From developer point of view you are absolutely right and all makes sense. But for the end user I think it would be more convenient when you don't have to
1. close the article edit view
2. switch to the associations component
3. select the type
4. select the article
5. select the language
6. click on the article
7. select the target language

before the compare view opens.

Mostly you want to do that when you edit the article directly. Probably a button beside the "Save as copy" button would make sense to switch to the compare view. But I'm more of a developer than UX person so I think would be nice to have some input from UX people as well.

**Comment by andrepereiradasilva** [comment]
> This is a very handy functionality, but should this not being built into the article manager itself, when editing an article? On a first glance it will be difficult to understand why the compare editor is on a completely different place, despite that the association is managed in the article itself.

@laoneo IMHO, if this componet exist it would be relatively straigh forward to integrate it in the "Associations" Tab "Edit" buttons (or something like it) in the future (excluding ACL checks, basicly should be opening the modal with the com_associations URI, instead of edit URI). Here:
image

**Comment by mbabker** [comment]
If I'm not mistaken you'd still be able to set associations from within the article edit view, it just seems like this mainly creates an alternative workflow to set associations and a way to have the content from two associated items open side-by-side instead of needing modals or separate windows.  Just from the screenshots and what little I know of associations since I get chastised whenever I touch content on the one multilingual site I manage I don't see any UX issues with this at the moment.

**Comment by infograf768** [comment]
It is not only extremely handy to use it when you create a multilingual site: prepare your site in one language and then do all associations (it includes also create a new one or change the assoc to another item) from the original reference item, but also, alter on when you change some stuff in the "base" item and want to modify the assoc content (what we call the Target item).

**Comment by infograf768** [comment]
It took me a few minutes to make an association helper for sermonspeaker:
screen shot 2017-01-10 at 19 03 51

@Bakual 
[associations.php.zip](https://github.com/joomla/joomla-cms/files/697014/associations.php.zip)
(Still have an error in side by side with this file though)

+++ a new ini string
COM_SERMONSPEAKER_CATEGORIES="Categories"

File: administrator/components/com_associations/helpers/associations.php [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+	public static function getExtensionHelper($extensionName)
+	{
+		if (! self::hasSupport($extensionName))
+		{
+			return null;
+		}
+
+		$support = self::$extensionsSupport[$extensionName];
+
+		return $support->get('helper');
```
Comment (zero-24) [inline administrator/components/com_associations/helpers/associations.php]: I'm not 100% sure can't we change this to: `self::$extensionsSupport[$extensionName]->get('helper');`

File: administrator/components/com_associations/models/associations.php [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+			$query->select($db->qn($fields['access'], 'access'));
+
+			// Join over the access levels.
+			$query->select($db->qn('ag.title', 'access_level'))
+				->join('LEFT', $db->qn('#__viewlevels', 'ag') . ' ON ' . $db->qn('ag.id') . ' = ' . $db->qn($fields['access']));
+
+			$groupby[] = 'ag.title';
+
+			// Implement View Level Access.
+			if (!$user->authorise('core.admin', $extension))
```
Comment (zero-24) [inline administrator/components/com_associations/models/associations.php]: please double check the `$extension` should this not be `$extensionName`

File: administrator/components/com_associations/models/associations.php [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+			{
+				$query->where($fields['access'] . ' IN (' . implode(',', $user->getAuthorisedViewLevels()) . ')');
+			}
+		}
+
+		// If component item type is menus we need to remove the root item and the administrator menu.
+		if ($extension === 'com_menus')
```
Comment (zero-24) [inline administrator/components/com_associations/models/associations.php]: same here `$extensionName`?

File: administrator/components/com_associations/views/association/view.html.php [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+		$bar->appendButton(
+			'Custom', '<button onclick="Joomla.submitbutton(\'target\')"'
+			. 'class="btn btn-small btn-success"><span class="icon-apply icon-white"></span>'
+			. JText::_('COM_ASSOCIATIONS_SAVE_TARGET') . '</button>', 'target'
+		);
+
+		JToolBarHelper::custom('copy', 'copy.png', '', 'COM_ASSOCIATIONS_COPY_REFERENCE', false);
+
+		JToolbarHelper::cancel('association.cancel', 'JTOOLBAR_CLOSE');
+		JToolbarHelper::help('JGLOBAL_HELP');
```
Comment (zero-24) [inline administrator/components/com_associations/views/association/view.html.php]: I guess this needs to be a todo?

Comment (zero-24) [inline administrator/components/com_associations/views/association/view.html.php]: as we need to add a help page here?

File: administrator/components/com_associations/views/associations/view.html.php [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+				{
+					unset($this->activeFilters['category_id']);
+					$this->filterForm->removeField('category_id', 'filter');
+				}
+				if ($extension !== 'com_menu')
+				{
+					unset($this->activeFilters['menutype']);
+					$this->filterForm->removeField('menutype', 'filter');
+				}
+				if (! in_array($extension, array('com_categories', 'com_menus')))
```
Comment (zero-24) [inline administrator/components/com_associations/views/associations/view.html.php]: again extension to extensionname?

**Review by zero-24 — CHANGES_REQUESTED** [review]

File: administrator/components/com_associations/models/associations.php (hunk 3 above)
Comment (infograf768) [inline administrator/components/com_associations/models/associations.php]: Indeed, although that check is a bit useless as these items will never get a language assigned and therefore should never display in the assoc manager.

File: administrator/components/com_associations/views/associations/view.html.php (hunk 5 above)
Comment (infograf768) [inline administrator/components/com_associations/views/associations/view.html.php]: Not in at case imho.
We have above:
`$extension = AssociationsHelper::getSupportedExtension($extensionName);`

**Comment by infograf768** [comment]
@Bakual 
Corrected the error in the sermonspeaker helper
[associations.php.zip](https://github.com/joomla/joomla-cms/files/698480/associations.php.zip)

**Comment by Bakual** [comment]
@infograf768 Didn't have yet time t test anything here, but thanks. I've already added it to my repo :smile: 
https://github.com/Bakual/SermonSpeaker/blob/master/com_sermonspeaker/admin/helpers/associations.php

**Comment by infograf768** [comment]
👍  @Bakual 

Eagerly waiting for help from you to solve first the checked_out issue.

**Comment by AlexRed** [comment]
I have tested this item :white_check_mark: successfully on 7b0b75f2a02b690ef39cc3b3aa5096917cb795a8<br /><br />Ok for me.
But please consider to change the "Clear" button in the Target side in "Delete Associations" to avoid confusion ad help in the translation procedure.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13537">issues.joomla.org/tracker/joomla-cms/13537</a>.</sub>

**Comment by infograf768** [comment]
Here is a demo of the checked_out issue.
The item concerned has been checked out by another user.
The user editing has no com_checkin manage permission.

associationscheckedout

**Comment by Bakual** [comment]
I have tested this item :white_check_mark: successfully on 7b0b75f2a02b690ef39cc3b3aa5096917cb795a8<br /><br />Small thing:
In the "Select Target" modal, the tooltips for the search fields above are hidden behind the modal header. In other similar modals we set the tooltips to go below the field and not on top.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13537">issues.joomla.org/tracker/joomla-cms/13537</a>.</sub>

File: administrator/components/com_associations/views/association/view.html.php (hunk 4 above)
Comment (rdeutz) [inline administrator/components/com_associations/views/association/view.html.php]: yes is a todo

File: administrator/components/com_associations/views/associations/view.html.php (hunk 5 above)
Comment (rdeutz) [inline administrator/components/com_associations/views/associations/view.html.php]: extensionName would be correct

File: administrator/components/com_associations/helpers/associations.php (hunk 1 above)
Comment (rdeutz) [inline administrator/components/com_associations/helpers/associations.php]: should work but I wouldn't change it here

**Comment by infograf768** [comment]
For the Select Target Modal tooltip, change in `/administrator/components/com_associations/views/associations/tmpl/modal.php`
`JHtml::_('bootstrap.tooltip');`
to
`JHtml::_('bootstrap.tooltip', '.hasTooltip', array('placement' => 'bottom'));`

File: administrator/components/com_associations/views/associations/view.html.php [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+				{
+					unset($this->activeFilters['state']);
+					$this->filterForm->removeField('state', 'filter');
+				}
+				if ($type !== 'category')
+				{
+					unset($this->activeFilters['category_id']);
+					$this->filterForm->removeField('category_id', 'filter');
+				}
+				if ($extension !== 'com_menu')
```
Comment (andrepereiradasilva) [inline administrator/components/com_associations/views/associations/view.html.php]: `com_menus´?

Comment (infograf768) [inline administrator/components/com_associations/views/associations/view.html.php]: Agree, that is also a wrong one.

File: administrator/components/com_associations/views/associations/view.html.php (hunk 5 above)
Comment (infograf768) [inline administrator/components/com_associations/views/associations/view.html.php]: oops...

File: administrator/components/com_associations/associations.xml [hunk 7]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+<extension type="component" version="3.7" method="upgrade">
+	<name>com_associations</name>
+	<author>Joomla! Project</author>
+	<creationDate>Januar 2017</creationDate>
```
Comment (infograf768) [inline administrator/components/com_associations/associations.xml]: `January`

**Comment by infograf768** [comment]
Created helper for weblinks.
https://github.com/joomla-extensions/weblinks/pull/294

File: administrator/components/com_associations/views/associations/view.html.php [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+				{
+					unset($this->activeFilters['state']);
+					$this->filterForm->removeField('state', 'filter');
+				}
+				if ($type !== 'category')
+				{
+					unset($this->activeFilters['category_id']);
+					$this->filterForm->removeField('category_id', 'filter');
+				}
+				if ($extensionName !== 'com_menu')
```
Comment (infograf768) [inline administrator/components/com_associations/views/associations/view.html.php]: should be `com_menus`

Comment (infograf768) [inline administrator/components/com_associations/views/associations/view.html.php]: Once this is corrected, we do get the `Select Menu` filter
screen shot 2017-01-12 at 09 30 36

BUT when choosing a menu type, we get a 500
`500 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '= 'mainmenu-fr-fr' GROUP BY `a`.`id`,`a`.`title`,`a`.`language`,`l`.`title`,`l`.' at line 7 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '= 'mainmenu-fr-fr' GROUP BY `a`.`id`,`a`.`title`,`a`.`language`,`l`.`title`,`l`.' at line 7 `

Comment (infograf768) [inline administrator/components/com_associations/views/associations/view.html.php]: Found the culprit

Comment (infograf768) [inline administrator/components/com_associations/views/associations/view.html.php]: See patch https://github.com/rdeutz/joomla-cms/pull/8 correcting the menutype.
Still remains to change 'com_menu' to 'com_menus'

**Comment by infograf768** [comment]
@rdeutz 
These are ready, correcting com_menus typo, checked_out wrong variable, $cancheckin and menu helper
https://github.com/rdeutz/joomla-cms/pull/11
https://github.com/rdeutz/joomla-cms/pull/10
https://github.com/rdeutz/joomla-cms/pull/9
https://github.com/rdeutz/joomla-cms/pull/8

File: administrator/language/en-GB/en-GB.com_menus.ini [hunk 9]
```diff
@@ -145,13 +146,11 @@ COM_MENUS_MENU_MENUTYPE_LABEL="Menu Type"
...
 COM_MENUS_MENU_SEARCH_FILTER="Search in Title or Menu type"
 COM_MENUS_MENU_SPRINTF="Menu: %s"
-COM_MENUS_TYPE_SYSTEM="System Links"
+COM_MENUS_MENUS="Menu items"
 COM_MENUS_MENU_TITLE_DESC="The title of the menu to display in the Administrator Menubar and lists."
 COM_MENUS_MENU_TYPE_NOT_ALLOWED="This is a reserved menutype."
 COM_MENUS_MENUS_FILTER_SEARCH_DESC="Search in title and menu type."
 COM_MENUS_MENUS_FILTER_SEARCH_LABEL="Search Menus"
-COM_MENUS_PAGE_OPTIONS_LABEL="Page Display"
-COM_MENUS_LABEL_HIDDEN="Hidden"
```
Comment (andrepereiradasilva) [inline administrator/language/en-GB/en-GB.com_menus.ini]: why are there removed?

File: administrator/language/en-GB/en-GB.com_menus.ini [hunk 10]
```diff
@@ -145,13 +146,11 @@ COM_MENUS_MENU_MENUTYPE_LABEL="Menu Type"
 COM_MENUS_MENU_SAVE_SUCCESS="Menu successfully saved"
 COM_MENUS_MENU_SEARCH_FILTER="Search in Title or Menu type"
 COM_MENUS_MENU_SPRINTF="Menu: %s"
-COM_MENUS_TYPE_SYSTEM="System Links"
```
Comment (andrepereiradasilva) [inline administrator/language/en-GB/en-GB.com_menus.ini]: why are this removed?

File: administrator/components/com_associations/associations.php [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+// Check if user has permission to access the component item type.
+$itemtype = JFactory::getApplication()->input->get('itemtype', '', 'string');
+
+if ($itemtype !== '')
+{
+	list($extensionName, $typeName) = explode('.', $itemtype);
+
+	if (!AssociationsHelper::hasSupport($extensionName))
+	{
+		throw new Exception(JText::_('COM_ASSOCIATIONS_COMPONENT_NOT_SUPPORTED') . ' ' . $itemType->realcomponent, 404);
```
Comment (andrepereiradasilva) [inline administrator/components/com_associations/associations.php]: Does `$itemType->realcomponent` exists? or is a leftover?

File: administrator/language/en-GB/en-GB.com_menus.ini (hunk 9 above)
Comment (infograf768) [inline administrator/language/en-GB/en-GB.com_menus.ini]: Agree, they should not

File: administrator/language/en-GB/en-GB.com_menus.ini (hunk 10 above)
Comment (infograf768) [inline administrator/language/en-GB/en-GB.com_menus.ini]: Agree, they should not

File: administrator/components/com_associations/associations.php (hunk 11 above)
Comment (infograf768) [inline administrator/components/com_associations/associations.php]: Also, the lang string `COM_ASSOCIATIONS_COMPONENT_NOT_SUPPORTED` is not present in the ini file.

Comment (infograf768) [inline administrator/components/com_associations/associations.php]: Maybe we should have something like:
`throw new Exception(JText::sprintf('COM_ASSOCIATIONS_COMPONENT_NOT_SUPPORTED', JText::_($extensionName), JText::_($typeName), 404);`

and 
`COM_ASSOCIATIONS_COMPONENT_NOT_SUPPORTED= "The extension %1$s does not support multilingual associations for the item type %2$s."

File: administrator/language/en-GB/en-GB.com_menus.ini (hunk 10 above)
Comment (rdeutz) [inline administrator/language/en-GB/en-GB.com_menus.ini]: I didn't remove them I just fixed the alpha ordering

File: administrator/language/en-GB/en-GB.com_menus.ini (hunk 9 above)
Comment (rdeutz) [inline administrator/language/en-GB/en-GB.com_menus.ini]: I didn't remove them I just fixed the alpha ordering

File: administrator/components/com_associations/associations.php (hunk 11 above)
Comment (rdeutz) [inline administrator/components/com_associations/associations.php]: fixed with 406d18d2428a4ea11eb2e80f4f8c997a4cc621e1 thanks!

**Comment by andrepereiradasilva** [comment]
@rdeutz @infograf768 it seems this PR misses the update (extension + asset) and installation part (extension + asset).

**Comment by rdeutz** [comment]
@andrepereiradasilva yes I am working on it just at this moment :-)

File: installation/sql/mysql/joomla.sql [hunk 12]
```diff
@@ -1401,9 +1402,10 @@ INSERT INTO `#__menu` (`id`, `menutype`, `title`, `alias`, `note`, `path`, `link
...
+(20, 'main', 'com_tags', 'Tags', '', 'Tags', 'index.php?option=com_tags', 'component', 0, 1, 1, 29, 0, '0000-00-00 00:00:00', 0, 1, 'class:tags', 0, '', 39, 40, 0, '', 1),
+(21, 'main', 'com_postinstall', 'Post-installation messages', '', 'Post-installation messages', 'index.php?option=com_postinstall', 'component', 0, 1, 1, 32, 0, '0000-00-00 00:00:00', 0, 1, 'class:postinstall', 0, '', 41, 42, 0, '*', 1),
+(22, 'menu', 'com_associations', 'Multilingual Associations', '', 'Multilingual Associations', 'index.php?option=com_associations', 'component', 1, 1, 1, 28, 0, '0000-00-00 00:00:00', 0, 0, 'class:associations', 0, '', 37, 38, 0, '*', 1),
```
Comment (andrepereiradasilva) [inline installation/sql/mysql/joomla.sql]: the component id is 34, not 28, see above

File: installation/sql/postgresql/joomla.sql [hunk 13]
```diff
@@ -504,7 +504,8 @@ INSERT INTO "#__extensions" ("extension_id", "name", "type", "element", "folder"
 (30, 'com_contenthistory', 'component', 'com_contenthistory', '', 1, 1, 1, 0, '', '', '', '', 0, '1970-01-01 00:00:00', 0, 0),
 (31, 'com_ajax', 'component', 'com_ajax', '', 1, 1, 1, 1, '', '', '', '', 0, '1970-01-01 00:00:00', 0, 0),
 (32, 'com_postinstall', 'component', 'com_postinstall', '', 1, 1, 1, 1, '', '', '', '', 0, '1970-01-01 00:00:00', 0, 0),
-(33, 'com_fields', 'component', 'com_fields', '', 1, 1, 1, 1, '', '', '', '', 0, '1970-01-01 00:00:00', 0, 0);
+(33, 'com_fields', 'component', 'com_fields', '', 1, 1, 1, 1, '', '', '', '', 0, '1970-01-01 00:00:00', 0, 0),
+(33, 'com_associations', 'component', 'com_associations', '', 1, 1, 1, 1, '', '', '', '', 0, '1970-01-01 00:00:00', 0, 0);
```
Comment (andrepereiradasilva) [inline installation/sql/postgresql/joomla.sql]: should be id 34
missing asset

File: installation/sql/sqlazure/joomla.sql [hunk 14]
```diff
@@ -805,7 +805,9 @@ SELECT 31, 'com_ajax', 'component', 'com_ajax', '', 1, 1, 1, 1, '', '', '', '',
 UNION ALL
 SELECT 32, 'com_postinstall', 'component', 'com_postinstall', '', 1, 1, 1, 1, '', '', '', '', 0, '1900-01-01 00:00:00', 0, 0
 UNION ALL
-SELECT 33, 'com_fields', 'component', 'com_fields', '', 1, 1, 1, 0, '', '', '', '', 0, '1900-01-01 00:00:00', 0, 0;
+SELECT 33, 'com_fields', 'component', 'com_fields', '', 1, 1, 1, 0, '', '', '', '', 0, '1900-01-01 00:00:00', 0, 0
+UNION ALL
+SELECT 34, 'com_associations', 'component', 'com_associations', '', 1, 1, 1, 0, '', '', '', '', 0, '1900-01-01 00:00:00', 0, 0;
```
Comment (andrepereiradasilva) [inline installation/sql/sqlazure/joomla.sql]: missing asset

File: installation/sql/mysql/joomla.sql (hunk 12 above)
Comment (rdeutz) [inline installation/sql/mysql/joomla.sql]: ups wrong file version

File: administrator/components/com_admin/script.php [hunk 15]
```diff
@@ -285,6 +285,7 @@ protected function updateManifestCaches()
 			array('component', 'com_postinstall', '', 1),
 			array('component', 'com_joomlaupdate', '', 1),
 			array('component', 'com_fields', '', 1),
+            array('component', 'com_associations', '', 1),
```
Comment (infograf768) [inline administrator/components/com_admin/script.php]: should be tabs instead of spaces and align to the others

Comment (wilsonge) [inline administrator/components/com_admin/script.php]: Done

**Comment by wilsonge** [comment]
Good work to everyone involved. Especially @rdeutz @infograf768 and especially @jreys for getting this started with your gsoc work :)

**Comment by puneet0191** [comment]
Thanks guys!  @rdeutz @infograf768 @alikon @andrepereiradasilva and @jreys :+1: and all the testers as well

**Comment by sloanthrasher** [comment]
Will the side-by-side view be available when editing an article on the front-end?

**Comment by infograf768** [comment]
> Will the side-by-side view be available when editing an article on the front-end?

Not in 3.7.0
I am not aware of anybody working on a frontend Interface.

**Comment by joomleb** [comment]
Hi guys, 
to really be able to manage sites multilanguage sites (quickly and simply) I think 3 things are a must:
1 - A fron-end interface, on normal sites is unthinkable to give translators an administrator access
2 - A layout more simple as possible. This is way I want put in evidence the [Speedy Translate extension](https://www.solidres.com/subscribe/speedy-translate) by Solidres. They found some solutions that in my opinion should be a need on joomla Multilanguage. In first of all I want put in evidence the "Inline Translation" feature. Click on a flag is 1 click, while click on a selector, choose, and select are "3 clicks".
3 - Last but not least, an article a lot of time have to be ready quaickly and it is hard to have all translated well soon. So, a "pre" - "auto-populate" button can be a solution (where clicking all the enabled languages are populated into the articles), as it is running (one by one) the [Translater CKEditor plugin](http://ckeditor.com/addon/xdsofttranslater) .

Many Thanks to all for this great new feature...

**Comment by Bakual** [comment]
@joomleb This is a closed (merged) PR and thus doesn't get much attention. You would have to create a new issue if you have found a real issue. If you just need assistance with this new feature, feel free to use the forums (http://forum.joomla.org/)

**Pull request merged**